File: bench/python/bench.py
#!/usr/bin/env python3
"""Python CSV benchmark: read 100k rows and aggregate by category."""

import csv
import sys
import time
from collections import defaultdict
from pathlib import Path

DATA_FILE = Path(__file__).parent.parent.parent / "data" / "test.csv"


def run() -> None:
    totals: dict[str, dict] = defaultdict(lambda: {
        "revenue": 0.0,
        "count": 0,
        "value_sum": 0.0,
        "value_max": float("-inf"),
        "value_min": float("inf"),
    })

    with DATA_FILE.open(newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            cat = row["category"]
            value = float(row["value"])
            qty = int(row["quantity"])
            g = totals[cat]
            g["revenue"] += value * qty
            g["count"] += 1
            g["value_sum"] += value
            if value > g["value_max"]:
                g["value_max"] = value
            if value < g["value_min"]:
                g["value_min"] = value

    print(f"{'Category':<12} {'Count':>8} {'Revenue':>14} {'Avg Value':>12} {'Max':>8} {'Min':>8}")
    print("-" * 66)
    for cat in sorted(totals):
        g = totals[cat]
        avg = g["value_sum"] / g["count"]
        print(f"{cat:<12} {g['count']:>8,} {g['revenue']:>14,.2f} {avg:>12.2f} {g['value_max']:>8.2f} {g['value_min']:>8.2f}")


def main() -> None:
    start = time.perf_counter()
    run()
    elapsed = time.perf_counter() - start
    print(f"\nElapsed: {elapsed:.4f}s")


if __name__ == "__main__":
    main()

File: bench/go/bench.go
package main

import (
	"encoding/csv"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"time"
)

type Group struct {
	revenue  float64
	count    int
	valueSum float64
	valueMax float64
	valueMin float64
}

func main() {
	_, thisFile, _, _ := runtime.Caller(0)
	dataFile := filepath.Join(filepath.Dir(thisFile), "..", "..", "data", "test.csv")

	f, err := os.Open(dataFile)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	defer f.Close()

	start := time.Now()

	r := csv.NewReader(f)
	header, err := r.Read()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	colIndex := map[string]int{}
	for i, col := range header {
		colIndex[col] = i
	}

	totals := map[string]*Group{}

	for {
		row, err := r.Read()
		if err != nil {
			break
		}
		cat := row[colIndex["category"]]
		value, _ := strconv.ParseFloat(row[colIndex["value"]], 64)
		qty, _ := strconv.Atoi(row[colIndex["quantity"]])

		g, ok := totals[cat]
		if !ok {
			g = &Group{valueMax: math.Inf(-1), valueMin: math.Inf(1)}
			totals[cat] = g
		}
		g.revenue += value * float64(qty)
		g.count++
		g.valueSum += value
		if value > g.valueMax {
			g.valueMax = value
		}
		if value < g.valueMin {
			g.valueMin = value
		}
	}

	cats := make([]string, 0, len(totals))
	for cat := range totals {
		cats = append(cats, cat)
	}
	sort.Strings(cats)

	fmt.Printf("%-12s %8s %14s %12s %8s %8s\n", "Category", "Count", "Revenue", "Avg Value", "Max", "Min")
	for i := 0; i < 66; i++ {
		fmt.Print("-")
	}
	fmt.Println()
	for _, cat := range cats {
		g := totals[cat]
		avg := g.valueSum / float64(g.count)
		fmt.Printf("%-12s %8d %14.2f %12.2f %8.2f %8.2f\n", cat, g.count, g.revenue, avg, g.valueMax, g.valueMin)
	}

	elapsed := time.Since(start)
	fmt.Printf("\nElapsed: %.4fs\n", elapsed.Seconds())
}
